bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java

```diff
@@ -280,30 +256,20 @@ private void modelBlockReceived(ModbusRegisterArray registers) {
      * @param block
      */
     private void readCommonBlock(ModelBlock block) {
-        BasicModbusReadRequestBlueprint request = new BasicModbusReadRequestBlueprint(slaveId,
+        ModbusReadRequestBlueprint request = new ModbusReadRequestBlueprint(slaveId,
                 ModbusReadFunctionCode.READ_MULTIPLE_REGISTERS, block.address, // Start address
                 block.length, // number or words to return
                 maxTries);
 
-        PollTask task = new BasicPollTaskImpl(endpoint, request, new ModbusReadCallback() {
-
-            @Override
-            public void onRegisters(ModbusReadRequestBlueprint request, ModbusRegisterArray registers) {
-                parseCommonBlock(registers);
-            }
-
-            @Override
-            public void onError(ModbusReadRequestBlueprint request, Exception error) {
+        comms.submitOneTimePoll(request, result -> {
+            if (result.hasError()) {
+                Exception error = (@NonNull Exception) result.getCause();
```

Comment:
Why not just have `getCause()` return an optional so you can use it to replace `hasError` while also avoiding all of the `@NonNull` casting you are doing?

Reply:
Makes sense actually...👍